Mathlib/Analysis/Calculus/Gradient/Basic.lean

```diff
@@ -0,0 +1,321 @@
+
+theorem HasGradientAt.gradient (h : HasGradientAt f f' x) : ∇ f x = f' :=
+  h.differentiableAt.hasGradientAt.unique h
+
+theorem gradient_eq {f' : F → F} (h : ∀ x, HasGradientAt f (f' x) x) : ∇ f = f' :=
+  funext fun x => (h x).gradient
+
+variable {g : 𝕜 → 𝕜} {g' u : 𝕜} {L' : Filter 𝕜}
+
+theorem toDual_eq_StarRingEnd : ((toDual 𝕜 𝕜) g') 1 = starRingEnd 𝕜 g' := by simp
+
+theorem StarRingEnd_eq_toDual : ((toDual 𝕜 𝕜) (starRingEnd 𝕜 g')) 1 = g' := by simp
+
+theorem Mul_one_eq_SterRingEnd (g' : 𝕜) : ContinuousLinearMap.smulRight (1 : 𝕜 →L[𝕜] 𝕜)
+    (starRingEnd 𝕜 g') = (toDual 𝕜 𝕜) g' := by
+  refine Iff.mpr ContinuousLinearMap.ext_iff ?_
+  simp; intro v; rw [toDual_apply, IsROrC.inner_apply, mul_comm]
+
+theorem SterRingEnd_eq_Mul_one (g' : 𝕜) : ContinuousLinearMap.smulRight (1 : 𝕜 →L[𝕜] 𝕜)
+    g' = (toDual 𝕜 𝕜) (starRingEnd 𝕜 g') := by
+  refine Iff.mpr ContinuousLinearMap.ext_iff ?_
+  simp; intro; rw [toDual_apply, IsROrC.inner_apply, mul_comm]
+  rw [RingHomCompTriple.comp_apply, RingHom.id_apply]
+
+theorem HasGradientAtFilter.hasDerivAtFilter (h : HasGradientAtFilter g g' u L') :
+    HasDerivAtFilter g (starRingEnd 𝕜 g') u L' := by
+  rw [HasDerivAtFilter, Mul_one_eq_SterRingEnd]; exact h
+
+theorem HasDerivAtFilter.hasGradientAtFilter (h : HasDerivAtFilter g g' u L') :
+    HasGradientAtFilter g (starRingEnd 𝕜 g') u L' := by
+  rw [HasGradientAtFilter, ← SterRingEnd_eq_Mul_one]; exact h
+
+theorem HasGradientAt.hasDerivAt (h : HasGradientAt g g' u) :
+    HasDerivAt g (starRingEnd 𝕜 g') u := by
+  rw [HasGradientAt_iff_HasFDerivAt, hasFDerivAt_iff_hasDerivAt, toDual_eq_StarRingEnd] at h
+  exact h
+
+theorem HasGradientAt.hasDerivAt' {g : ℝ → ℝ} {g' u: ℝ} (h :HasGradientAt g g' u) :
+    HasDerivAt g g' u := h.hasDerivAt
+
+theorem HasDerivAt.hasGradientAt (h : HasDerivAt g g' u) :
+    HasGradientAt g (starRingEnd 𝕜 g') u := by
+  rw [HasGradientAt_iff_HasFDerivAt, hasFDerivAt_iff_hasDerivAt, StarRingEnd_eq_toDual]
+  exact h
+
+theorem HasDerivAt.hasGradientAt' {g : ℝ → ℝ} {g' u: ℝ} (h :HasDerivAt g g' u) :
+    HasGradientAt g g' u := h.hasGradientAt
+
+theorem gradient_deriv : ∇ g u = starRingEnd 𝕜 (deriv g u) := by
+  by_cases h: DifferentiableAt 𝕜 g u
+  · rw [h.hasGradientAt.hasDerivAt.deriv]
+    exact Eq.symm (IsROrC.conj_conj (∇ g u))
+  · rw [Gradient_zero_of_not_differentiableAt h, deriv_zero_of_not_differentiableAt h]
+    exact Eq.symm (RingHom.map_zero (starRingEnd 𝕜))
+
+theorem gradient_deriv' {g : ℝ → ℝ} {u: ℝ} : ∇ g u = deriv g u := gradient_deriv
+
+open Filter
+
+/-- If a function f has a derivative f' at x, a rescaled version of f around x converges to f',
+i.e., `n (f (x + (1/n) v) - f x)` converges to `f' v`. More generally, if `c n` tends to infinity
+and `c n * d n` tends to `v`, then `c n * (f (x + d n) - f x)` tends to `< f', v >`. This lemma
+expresses this fact, for functions having a derivative within a set. Its specific formulation is
+useful for tangent cone related discussions. -/
```

Comment:
The docstring does not really match the statement, as it doesn't talk about gradients.

Do you really need this lemma? Otherwise, I would just drop it as it is pretty specific and was introduced for the derivative only as a device to prove another result.

Reply:
Thank you for your advice! It seems that this theorem actually should not belong to this file. We think it is useful, but it just need not to place here. So we decide to delete it here and add at the place where we need the theorem.